### PR TITLE
multi-arch: set ARCHITECTURE equal to BUILD_ARCH

### PR DESCRIPTION
**Pull request by zhlhahaha**
Make bazel build for correct Architecture, when there is only one parameter in BUILD_ARCH

As we set `BUILD_ARCH=crossbuild-aarch64` in the build process, and the parameter does not pass to `bazel build` which lead to cross build not works at the moment.
https://github.com/kubevirt/project-infra/blob/39e4c9854a682d00f762868f58ec0b9a4bf6e013/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-periodics.yaml#L322-L323

**Special notes for your reviewer**:
This is a fix to https://github.com/kubevirt/kubevirt/issues/3558#issuecomment-1533261386

```release-note
NONE
```

### CODE REVIEW

**Comment by zhlhahaha** [comment]
cc: @rmohr @rthallisey

**Comment by andreabolognani** [comment]
The change looks fine, but isn't `BUILD_ARCH` being used inconsistently?

The prow job sets

```sh
BUILD_ARCH="crossbuild-aarch64"
DOCKER_TAG="<timestamp>-arm64"
```

but my reading of the script is that it expects, for multi-arch builds, something like

```sh
BUILD_ARCH="amd64,arm64"
DOCKER_TAG="<timestamp>"
```

instead.

So I think we might need something like

```diff
diff --git a/hack/multi-arch.sh b/hack/multi-arch.sh
index 7ede30961..e5bd47d20 100755
--- a/hack/multi-arch.sh
+++ b/hack/multi-arch.sh
@@ -19,13 +19,24 @@
 
 COMMAND=$1
 
+arch2tag() {
+    case "$1" in
+    *aarch64)
+        echo "arm64"
+        ;;
+    *x86_64)
+        echo "amd64"
+        ;;
+    esac
+}
+
 build_count=$(echo ${BUILD_ARCH//,/ } | wc -w)
 
 # Only add the tailing $arch when doing a multi-arch build
 if [ "$build_count" -gt 1 ]; then
     for arch in ${BUILD_ARCH//,/ }; do
         echo "[INFO] -- working on $arch --"
-        DOCKER_TAG=$DOCKER_TAG-$arch ARCHITECTURE=$arch hack/bazel-${COMMAND}.sh
+        DOCKER_TAG=$DOCKER_TAG-$(arch2tag ${arch}) ARCHITECTURE=$arch hack/bazel-${COMMAND}.sh
     done
 else
     ARCHITECTURE=${BUILD_ARCH} hack/bazel-${COMMAND}.sh
```

(completely untested) to make it possible to actually perform multi-arch builds by setting

```sh
BUILD_ARCH="x86_64,crossbuild-aarch64"
DOCKER_TAG="<timestamp>"
```

Am I missing something?

**Comment by zhlhahaha** [comment]
Hi, @andreabolognani 

When we do crossbuild, we need to set the BUILD_ARCH="crossbuild-aarch64". So if we build multi-arch image on a x86_64, we need to set `BUILD_ARCH="x86_64,crossbuild-aarch64"`

I have verified it, as you can see in the following comment: https://github.com/kubevirt/kubevirt/pull/8989#issuecomment-1371884282

**Comment by zhlhahaha** [comment]
So if we want to do both native build and cross build when setting `BUILD_ARCH="amd64,arm64"`, we may need another function to detect if the current CPU arch march with BUILD_ARCH, if not we set the `ARCHITECTURE` to `crossbuild-${BUILD_ARCH}`

However, the current work flow should be works fine without the previous amend. And we do not publish multi-arch image for now as Arm64 release is still experimental.

**Comment by zhlhahaha** [comment]
Hi @andreabolognani 
I have updated the PR. Now, it is supposed to work when we set `BUILD_ARCH="amd64,arm64"`. I am not sure if I miss understand your idea.

BTW, as you can see in `hack/push-container-manifest.sh`, when creating a multi-arch manifest, digests are used instead of tags. As a result, the specific tag assigned to the multi-arch images is not significant.

**Comment by andreabolognani** [comment]
> BTW, as you can see in `hack/push-container-manifest.sh`, when creating a multi-arch manifest, digests are used instead of tags. As a result, the specific tag assigned to the multi-arch images is not significant.

Good point, I hadn't realized that was the case.

> However, the current work flow should be works fine without the previous amend. And we do not publish multi-arch image for now as Arm64 release is still experimental.

Fair enough.

> I have updated the PR. Now, it is supposed to work when we set BUILD_ARCH="amd64,arm64". I am not sure if I miss understand your idea.

Sort of! I actually didn't mind the fact that you had to specify "crossbuild-aarch64" as opposed to "aarch64", but your new version dealing with that automatically is an improvement over the status quo of course.

I was mostly calling out the inconsistency in tag names, which for a multi-arch build will be

```
${DOCKER_TAG}-x86_64
${DOCKER_TAG}-crossbuild-aarch64
```

while the current aarch64 CI job publishes the result as

```
${DOCKER_TAG}-arm64
```

and generally I would expect this latter convention to be used for container images, as docker/podman use Go-style architecture names everywhere.

File: hack/multi-arch.sh [hunk 1]
```diff
@@ -18,15 +18,46 @@
...
+# it is consistent with the platform name specified in ../.bazelrc
+function format_archname() {
+    local local_platform=$(uname -m)
+    local platform=$1
+
+    if [ $# -lt 1 ]; then
+        echo ${local_platform}
+    else
+        case ${platform} in
+        x86_64* | i?86_64* | amd64*)
```
Comment (andreabolognani) [inline hack/multi-arch.sh:31]: Are the `*`s needed here? And what is `i?86_64*` supposed to match?

Comment (zhlhahaha) [reply hack/multi-arch.sh]: Yes, you are right.

**Comment by zhlhahaha** [comment]
>  generally I would expect this latter convention to be used for container images, as docker/podman use Go-style architecture names everywhere.

I have updated the PR, and use GO-style architecture names in image tags

**Comment by zhlhahaha** [comment]
@rthallisey Any future comments on the PR?

**Comment by zhlhahaha** [comment]
> /approve

Thanks Roman

**Pull request merged**